concordium-consensus/tests/scheduler/SchedulerTests/Helpers.hs

```diff
@@ -1,24 +1,430 @@
+    TestConfig ->
+    PersistentBSM pv (BS.HashedPersistentBlockState pv) ->
+    (BS.PersistentBlockState pv -> PersistentBSM pv a) ->
+    Types.GroupedTransactions ->
+    IO (SchedulerResult, a)
+runSchedulerTest config constructState extractor transactions = runTestBlockState computation
+  where
+    computation :: PersistentBSM pv (SchedulerResult, a)
+    computation = do
+        blockStateBefore <- constructState
+        (result, blockStateAfter) <- runScheduler config blockStateBefore transactions
+        (result,) <$> extractor blockStateAfter
+
+-- | Run the scheduler on transactions in a test environment.
+-- Allows for a block state monad computation for constructing the initial block state and takes a
+-- block state monad computation for extracting relevant values.
+--
+-- The transactions are provided using TransactionJSON.
+runSchedulerTestTransactionJson ::
+    forall pv a.
+    (Types.IsProtocolVersion pv) =>
+    TestConfig ->
+    PersistentBSM pv (BS.HashedPersistentBlockState pv) ->
+    (BS.PersistentBlockState pv -> PersistentBSM pv a) ->
+    [SchedTest.TransactionJSON] ->
+    IO (SchedulerResult, a)
+runSchedulerTestTransactionJson config constructState extractor transactionJsonList = do
+    transactions <- SchedTest.processUngroupedTransactions transactionJsonList
+    runSchedulerTest config constructState extractor transactions
+
+-- | Intermediate results collected while running a number of transactions.
+type IntermediateResults a = [(SchedulerResult, a)]
+
+-- | Run the scheduler on transactions in a test environment, while collecting all of the
+-- intermediate results and extracted values.
+runSchedulerTestWithIntermediateStates ::
+    forall pv a.
+    (Types.IsProtocolVersion pv) =>
+    TestConfig ->
+    PersistentBSM pv (BS.HashedPersistentBlockState pv) ->
+    (BS.PersistentBlockState pv -> PersistentBSM pv a) ->
+    Types.GroupedTransactions ->
+    IO (IntermediateResults a, BS.HashedPersistentBlockState pv)
+runSchedulerTestWithIntermediateStates config constructState extractor transactions =
+    runTestBlockState blockStateComputation
+  where
+    blockStateComputation :: PersistentBSM pv (IntermediateResults a, BS.HashedPersistentBlockState pv)
+    blockStateComputation = do
+        blockStateBefore <- constructState
+        foldM transactionRunner ([], blockStateBefore) transactions
+
+    transactionRunner ::
+        (IntermediateResults a, BS.HashedPersistentBlockState pv) ->
+        Types.TransactionGroup ->
+        PersistentBSM pv (IntermediateResults a, BS.HashedPersistentBlockState pv)
+    transactionRunner (acc, currentState) tx = do
+        (result, updatedState) <- runScheduler config currentState [tx]
+        extracted <- extractor updatedState
+        nextState <- BS.freezeBlockState updatedState
+        return (acc ++ [(result, extracted)], nextState)
+
+-- | Save and load block state, used to test the block state written to disc.
```

Comment:
Please add how this is recommended to be used (i.e., composed with the test function provided to one of the `run*Test` functions.